File: solutions/go/luhn/1/luhn.go
package luhn

import (
	"strings"
	"strconv"
)

func Valid(s string) bool {
	doublCheck := 1
	sum := 0
	s = strings.Replace(s, " ", "", -1)

	for i := len(s); i > 0; i-- {
		value := int(s[i-1])

		if value > 57 || value < 48 {
			return false
		}

		digit, _ := strconv.Atoi(string(value))
		if doublCheck%2 == 0 {
			digit = digit * 2
			if digit > 9 {
				digit -= 9
			}
		}
		sum += digit
		doublCheck++
	}
	if sum%10 == 0 && len(s) > 1 {
		return true
	}
	return false
}

File: solutions/ruby/luhn/1/luhn.rb
module Luhn
    def self.valid?(input)
        input.gsub!(/\s/, '')

        return false if input.length < 2 || input.match(/[^0-9]/)

        numbers = []
        input.reverse.chars.each_with_index do |num, i| 
            num = num.to_i
            if i % 2 == 0
                num
            else
                num = num * 2
                num = num > 9 ? num-9 : num
            end
            numbers.push num
        end
        numbers.sum % 10 == 0
    end
end
